### ISSUE TITLE
Multi-architecture wheels from the same source

### ISSUE THREAD

**Issue by hgaiser** [python-poetry/poetry#9445]
### Description

The project I'm working on is running on both ARM and x86 architectures. There is a dependency which contains architecture specific libraries, so there are different wheels for the two supported architectures hosted in a private repository (cloudsmith.io, using the `simple` interface).

If I run `poetry lock --no-update` on x86, the lock file contains only the x86 wheels. If I run the same on `arm64` it finds both the x86 wheels **and** the arm64 wheels. My suspicion is that poetry first finds the x86 wheels, and then then arm64 wheels. If it is running on an x86 architecture, it fulfills dependency after finding the x86 wheels. On arm64 architecture it sees the x86 wheel files and proceeds to find the arm64 files, but logging both in the poetry.lock file (as expected).

This is a minor nuisance because it means the lock file changes depending on which architecture you are running.

Related issues:

https://github.com/python-poetry/poetry/issues/9156
https://github.com/python-poetry/poetry/issues/5205

I tried the following configuration:

```
markupsafe = [
    {version="^2.1.0", markers="platform_machine!='x86_64'", source='piwheels'},
    {version="^2.1.0", markers="platform_machine=='x86_64'", source="pypi"}
]
```

However there was no change in the described outcome.

### Workarounds

Not really a workaround, aside from ignoring changes to poetry.lock.

### Poetry Installation Method

system package manager (eg: dnf, apt etc.)

### Operating System

Arch Linux (and Debian Bullseye on arm64)

### Poetry Version

1.8.3

### Poetry Configuration

```bash session
cache-dir = "/home/hgaiser/.cache/pypoetry"
experimental.new-installer = false
experimental.system-git-client = false
installer.max-workers = null
installer.modern-installation = true
installer.no-binary = null
installer.parallel = true
keyring.enabled = true
repositories.<private-source>.url = "https://dl.cloudsmith.io/basic/<private-source>/python/python/simple"
solver.lazy-wheel = true
virtualenvs.create = true
virtualenvs.in-project = true
virtualenvs.options.always-copy = false
virtualenvs.options.no-pip = false
virtualenvs.options.no-setuptools = false
virtualenvs.options.system-site-packages = false
virtualenvs.path = "{cache-dir}/virtualenvs"  # /home/hgaiser/.cache/pypoetry/virtualenvs
virtualenvs.prefer-active-python = false
virtualenvs.prompt = "{project_name}-py{python_version}"
warnings.export = true
```
```


### Python Sysconfig

_No response_

### Example pyproject.toml

_No response_

### Poetry Runtime Logs

```bash session
N/A
```

**Comment by dimbleby** [comment]
`poetry lock` does not care what platform you are running on, pretty sure you have misunderstood something.  Perhaps you just need to clear your cache.

**Comment by hgaiser** [comment]
I agree, I don't think the issue is specific to an architecture, but it is occurring because I am using different architectures. My guess is, but I'm trying to verify this, that the logic is as follows:

1. Iterate over the wheels to find a matching dependency.
2. At the first matching dependency, exit, ignoring any additional wheels that exist from a repository.

Meaning if the x86 wheel is found first, the arm64 wheel is never reported and never added to poetry.lock.

**Comment by dimbleby** [comment]
what you wrote is

> If I run poetry lock --no-update on x86, the lock file contains only the x86 wheels

but I assert that `poetry lock` does not care that you are on x86, therefore you have misunderstood something.  

Your guess is wrong.  The lockfile should be expected to contain all available distributions, regardless of the platform on which you performed the locking.

If this is not the case, please provide a way to reproduce.

**Comment by hgaiser** [comment]
I know what I wrote, I am describing what I observed. I'm not pretending I am fully knowledgeable of poetry's inner workings so the chances that I misunderstood are not unlikely, but your messages seem quite confrontational, not sure if that is as intended. If I did something wrong, let me know.

> Your guess is wrong. The lockfile should be expected to contain all available distributions, regardless of the platform on which you performed the locking.
> 
> If this is not the case, please provide a way to reproduce.

Again, this is not what I observe. I will try to setup a minimum reproducible example. I cannot grant access to the project I'm working on since it is on a private repository. I will try to setup a simple "private" repository.

I noticed that wheels from pypi seem to be logged correctly in poetry.lock (meaning both arm64 and x86 wheels are listed). Perhaps the issue is limited to the HTTP repository format.

**Comment by dimbleby** [comment]
I am trying to encourage you to looking for a explanation other than "locking on one architecture gives a different answer than locking on another" - because I am confident that the locking process does not care what architecture you are running on.

eg hence my suggestion that the cache might be relevant.
